### PR TITLE
Fixing bug in log file purge logic (#393)

### PR DESCRIPTION
**Pull request by mathewc**
In the new purge logic we added, we discovered a bug where if the function log directory does not exist (e.g if no functions have ever been added or ran) the purge logic fails on startup.

### CODE REVIEW

File: src/WebJobs.Script.WebHost/SecretManager.cs [hunk 1]
```diff
@@ -109,41 +109,53 @@ public virtual FunctionSecrets GetFunctionSecrets(string functionName)
...
+                // that we don't delete logs/secrets for functions that transition into/out of
+                // invalid unparsable states.
+                var functionLookup = Directory.EnumerateDirectories(rootScriptPath).ToLookup(p => Path.GetFileName(p), StringComparer.OrdinalIgnoreCase);
 
-            var secretsDirectory = new DirectoryInfo(_secretsPath);
-            foreach (var secretFile in secretsDirectory.GetFiles("*.json"))
-            {
-                if (string.Compare(secretFile.Name, ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0)
+                var secretsDirectory = new DirectoryInfo(_secretsPath);
+                if (!Directory.Exists(_secretsPath))
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/SecretManager.cs]: I added this not exists check, and the rest of the changes are me pushing the entire method into a try/catch

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 2]
```diff
@@ -235,36 +235,48 @@ protected virtual void Initialize()
...
+                // invalid unparsable states.
+                var functionLookup = Directory.EnumerateDirectories(this.ScriptConfig.RootScriptPath).ToLookup(p => Path.GetFileName(p), StringComparer.OrdinalIgnoreCase);
 
-            string rootLogFilePath = Path.Combine(this.ScriptConfig.RootLogPath, "Function");
-            var logFileDirectory = new DirectoryInfo(rootLogFilePath);
-            foreach (var logDir in logFileDirectory.GetDirectories())
-            {
-                if (!functionLookup.Contains(logDir.Name))
+                string rootLogFilePath = Path.Combine(this.ScriptConfig.RootLogPath, "Function");
+                if (!Directory.Exists(rootLogFilePath))
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Same here: added this not exists check, and wrapped everything in a try/catch

File: test/WebJobs.Script.Tests/ScriptHostManagerTests.cs [hunk 3]
```diff
@@ -129,6 +130,45 @@ public void RunAndBlock_DisposesOfHost_WhenExceptionIsThrown()
...
+            {
+                RootScriptPath = functionDir,
+                RootLogPath = logDir,
+                FileLoggingEnabled = true
+            };
+            ScriptHostManager hostManager = new ScriptHostManager(config);
+
+            Task runTask = Task.Run(() => hostManager.RunAndBlock());
+
+            await TestHelpers.Await(() => hostManager.IsRunning, timeout: 10000);
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/ScriptHostManagerTests.cs]: This test would fail before the fix

File: src/WebJobs.Script.WebHost/SecretManager.cs [hunk 4]
```diff
@@ -109,41 +109,53 @@ public virtual FunctionSecrets GetFunctionSecrets(string functionName)
...
+                            secretFile.Delete();
+                        }
+                        catch
+                        {
+                            // Purge is best effort
+                        }
                     }
                 }
             }
+            catch
```
Comment (davidebbo) [inline src/WebJobs.Script.WebHost/SecretManager.cs]: Not a big fan of blindly catching everything here. The only downside of the original issue is that it dumped an exception to the trace, which is a good thing to help find bugs. This try/catch would have masked the issue.

File: src/WebJobs.Script/Host/ScriptHostManager.cs [hunk 5]
```diff
@@ -212,6 +212,8 @@ public void Stop()
                 {
                     instance.Dispose();
                 }
+
+                IsRunning = false;
```
Comment (davidebbo) [inline src/WebJobs.Script/Host/ScriptHostManager.cs]: That seems unrelated?

Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHostManager.cs]: Yes, I discovered this in the test I wrote where I call Stop and I expect IsRunning to be false at that point (test needs to wait for the stop to finish). Good to fix, but doesn't really affect users currently.

File: src/WebJobs.Script.WebHost/SecretManager.cs (hunk 4 above)
Comment (mathewc) [inline src/WebJobs.Script.WebHost/SecretManager.cs]: The issue is a little worse than that - it actually causes a startup time failure that will cause the host to continually recycle, until a function is added.

I could add back some error output in the catch though, but we don't want to block the host from running successfully if we can't purge.

Comment (mathewc) [inline src/WebJobs.Script.WebHost/SecretManager.cs]: Added logging

**Comment by davidebbo** [comment]
:shipit:
